tree.go

```diff
@@ -22,33 +21,34 @@ type Tree interface {
 	Root() ([]byte, error)
 }
 
-var _ Tree = &DefaultTree{}
+// trees is a global map used for keeping track of registered tree constructors for JSON serialization
+// The keys of this map should be kebab cased. E.g. "default-tree"
+var trees = make(map[string]TreeConstructorFn)
+
+func RegisterTree(treeName string, treeConstructor TreeConstructorFn) error {
+	if trees[treeName] != nil {
+		return fmt.Errorf("%s already registered", treeName)
+	}
+	trees[treeName] = treeConstructor
 
-type DefaultTree struct {
-	*merkletree.Tree
-	leaves [][]byte
-	root   []byte
+	return nil
 }
 
-func NewDefaultTree(_ Axis, _ uint) Tree {
-	return &DefaultTree{
-		Tree:   merkletree.New(sha256.New()),
-		leaves: make([][]byte, 0, 128),
+func GetTreeConstructorFn(treeName string) (TreeConstructorFn, error) {
+	if trees[treeName] == nil {
+		return nil, fmt.Errorf("%s not registered yet", treeName)
 	}
-}
 
-func (d *DefaultTree) Push(data []byte) error {
-	// ignore the idx, as this implementation doesn't need that info
-	d.leaves = append(d.leaves, data)
-	return nil
+	return trees[treeName], nil
 }
 
-func (d *DefaultTree) Root() ([]byte, error) {
-	if d.root == nil {
-		for _, l := range d.leaves {
-			d.Tree.Push(l)
+func GetTreeNameFromConstructorFn(treeConstructor TreeConstructorFn) string {
```

Comment:
Privatise. It's temporary until we break EDS constructors. We don't want anyone to start using it.